File: 7_Function_Overriding.py




# Runtime Polymorphism -> Dynamic Binding | Overriding | Late Binding

# CompileTime Polymorphism-> Static Binding | Overloading | Early Binding


class Animal:
    def make_sound(self):
        print("Animal makes a sound")

class Dog(Animal):
    def make_sound(self):  # Overriding the method from the base class
        print("Dog barks")

# Creating objects
animal = Animal()
dog = Dog()

animal.make_sound()  # Outputs: Animal makes a sound
dog.make_sound()     # Outputs: Dog barks

File: 7_Function_Overriding.cpp
#include <iostream>
using namespace std;

class Animal {
public:
    virtual void makeSound() {  // Using 'virtual' to allow overriding
        cout << "Animal makes a sound" << endl;
    }
};

class Dog : public Animal {
public:
    void makeSound() override {  // Overriding the method from the base class. The override keyword (optional but good practice) specifies that makeSound is meant to override a base class method.
        cout << "Dog barks" << endl;
    }
};

int main() {
    Animal* animal = new Animal();
    Animal* dog = new Dog();

    animal->makeSound();  // Outputs: Animal makes a sound
    dog->makeSound();     // Outputs: Dog barks

    delete animal;
    delete dog;
    return 0;
}
